### DISCUSSION TITLE
How NB manages to step into/pause etc. in Maven dependencies

### DISCUSSION THREAD

**Discussion by ghost** [apache/netbeans#5275] [General]
If there is a set of separate Ant projects which are related only by creating a CLASSPATH dependency tree, and the root (dependency-wise only) project is run, it is quite natural (and very handy) that stepping into a dependency, pausing at a dependency breakpoint etc. works. The user is navigated between local Java sources without a problem.

It is less evident in case of Maven projects, but surprisingly NB is often capable of doing exactly that, even if jars are passed via a local repo as 1.0-SNAPSHOT.

I guess that it does not work for a jar downloaded from central maven repository even if sources are available, because NB has no idea that these sources are the same as those of a NB project on a local disk.

But still, it works somehow if the local user repo is concerned. My question 1 is: how this mechanism of associating a Maven dependency with a local NB project works and how far it goes?

A question 2: would it be possible to extend the mechanism to "build with dependencies"?

**Comment by mbien** [comment]
> I guess that it does not work for a jar downloaded from central maven repository even if sources are available, because NB has no idea that these sources are the same as those of a NB project on a local disk.

you should be able to step through maven dependencies if you download the sources for them. You can also manipulate the source roots with help of the `window -> debugging -> sources` window, but this is rarely necessary.

> But still, it works somehow if the local user repo is concerned. My question 1 is: how this mechanism of associating a Maven dependency with a local NB project works and how far it goes?

its really simple. If the group, artifact id and version matches and NB has access to the sources, you can step through them while debugging. If you install an artifact in your local maven repo and add sources it should work. If you have a maven project open with matching artifact id etc it should find that one too (if not, try a maven clean install).

> A question 2: would it be possible to extend the mechanism to "build with dependencies"?

not quite sure what this means

**Reply by ghost** [reply]
> If you install an artifact in your local maven repo and add sources it should work. If you have a maven project open with matching artifact id etc it should find that one too (if not, try a maven clean install).

Yes, that works and is great.

> 
> > A question 2: would it be possible to extend the mechanism to "build with dependencies"?
> 
> not quite sure what this means

I would like to recompile all dependencies which happen to be currently open projects recognized as a dependency using the artifact id etc. you just spoke about. Otherwise, to assure that everything I am working on now is up-to-date, I need to browse through all open projects, looking for dependencies and recompilling them "manually". If there are tenths of such projects, it is a cumbersome operation and risks that something is still forgotten.

I just tested an open maven dependency: NB is capable of stepping into it etc., but "Build with dependencies" in the root project builds only the root project and ignores that dependency entirely.

**Reply by mbien** [reply]
> I just tested an open maven dependency: NB is capable of stepping into it etc., but "Build with dependencies" in the root project builds only the root project and ignores that dependency entirely.

ah, now I get it :) I suppose this would be probably possible to implement - not sure how hard that would be though.
